botorch/utils/sampling.py

```diff
@@ -873,3 +874,79 @@ def sparse_to_dense_constraints(
         A[i, indices.long()] = coefficients
         b[i] = rhs
     return A, b
+
+
+def optimize_posterior_samples(
+    paths: SamplePath,
+    bounds: Tensor,
+    candidates: Optional[Tensor] = None,
+    raw_samples: Optional[int] = 1024,
+    num_restarts: Optional[int] = 20,
+    maxiter: int = 50,
+    lr_init: float = 1e-3,
+    maximize: bool = True,
+) -> Tuple[Tensor, Tensor]:
+    r"""Cheaply maximizes posterior samples by random querying followed by vanilla
+    gradient descent on the best num_restarts points.
+
+    Args:
+        paths: Random Fourier Feature-based sample paths from the GP  
+        bounds: The bounds on the search space.
+        candidates: A priori good candidates (typically previous design points)
+            which acts as extra initial guesses for the optimization routine.
+        raw_samples: The number of samples with which to query the samples initially.
+        num_restarts The number of gradient descent steps to use on each of the best 
+        found initial points.
+        maxiter: The maximal permitted number of gradient descent steps.
+        lr: The stepsize of the gradient descent steps.
+
+    Returns:
+        A two-element tuple containing:
+            - A: A `num_optima x batch_size x d`-dim tensor of optimal inputs x*.
+            - b: A `num_optima x batch_size x 1`-dim tensor of optimal outputs f*.
+    """
+    if maximize:
+        path_func = lambda x: paths.forward(x)
+    else:
+        path_func = lambda x: -paths.forward(x) 
+
+    candidate_set = unnormalize(SobolEngine(
+        dimension=bounds.shape[1], scramble=True).draw(raw_samples), bounds)
+    # TODO add spray points
+    # queries all samples on all candidates - output raw_samples * num_objectives * num_optima
+    candidate_queries = path_func(candidate_set)
+    num_optima = candidate_queries.shape[0]
+    batch_size = candidate_queries.shape[1] if candidate_queries.ndim == 3 else 1
+    argtop_candidates = candidate_queries.argsort(dim=-1, descending=True)[
+        ..., 0:num_restarts]
+
+    # These are used as masks when retrieving the argmaxes
+    row_indexer = torch.arange(num_optima * batch_size).to(torch.long)
+
+    X_top_candidates = candidate_set[argtop_candidates,
+                                     :].requires_grad_(requires_grad=True)
+
+    for iter_ in range(maxiter):
+        # Decaying the learing rate - empirically, this has been well-calibrated
+        lr = lr_init / (iter_ + 1)
+        per_sample_outputs = path_func(X_top_candidates)
+        grads = torch.autograd.grad(
+            per_sample_outputs, X_top_candidates, grad_outputs=torch.ones_like(per_sample_outputs))[0]
+        X_top_candidates = torch.clamp(
+            X_top_candidates + lr * grads, bounds[0], bounds[1])  # TODO fix bounds here
```

Comment:
The sample paths here are deterministic (and smooth) functions - rather using a (manual) first-order gradient scheme here, we could probably get much faster convergence by using a (quasi)-second order method such as L-BFGS here.

Reply:
Sure, will fix!

Reply:
@Balandat After giving this some more thought, it seems pretty tricky to implement this in batch with the set of tools available. Would a for-loop solution suffice?

Reply:
I feel like a for loop across each sample would be quite costly. Another option here would be to just define a function that sums the `per_sample_outputs` and then call L-BFGS-B on that (with the domain being `k` copies of the original domain, with `k` the number of sample paths). This is actually what we do for performing acquisition function optimization across random restarts: https://github.com/pytorch/botorch/blob/9302055f36981e3413e8dce08ae050bd701f5e7a/botorch/generation/gen.py#L175-L203

Reply:
Okay, I did try it some time ago with the torch L-BFGS but I must have done something wrong if it indeed works here. I'll explore it!